### ISSUE TITLE
Post File requires FileStream

### ISSUE THREAD

**Issue by skrusty** [Azure/autorest#2850]
Hi,

We have an API that takes a file from the customer, it also takes a file name etc. When we generate the code using Autorest it takes a Stream in the method, but tries to cast it to a FileStream. 

If the developer passes a MemoryStream, then the method fails at:

```c#
FileStream _fileAsFileStream = file as FileStream;
                if (_fileAsFileStream != null)
````

Obviously this is because the MemoryStream might be from some in memory/db byte array or something.

What's the best way around this please? Looking for advice or suggestions.

Thanks in advance,
Ben

**Comment by fearthecowboy** [comment]
It's only casting down to `FileStream`  so that it can try to set some things

```
    if (_fileContentAsFileStream != null)
                {
                    ContentDispositionHeaderValue _contentDispositionHeaderValue = new ContentDispositionHeaderValue("form-data");
                    _contentDispositionHeaderValue.Name = "fileContent";
                    _contentDispositionHeaderValue.FileName = _fileContentAsFileStream.Name;
                    _fileContent.Headers.ContentDisposition = _contentDispositionHeaderValue;
                }

**Comment by nielsvoigt** [comment]
@fearthecowboy No it doesn't work fine. The API controller will receive null as an IFormFile. Right now there is no way from the outside to fix that, if you don't pass a FileStream, it will fail and there is no way to construct a FileStream without actual disk access. Apparently at least .NET needs the additional data to construct a IFormFile in the Controller successfully. Setting the .FileName to "unknown" or anything but an empty string or null will work though. 

This works (the if was removed, instead the cast will only influence the filename):

                StreamContent _file = new StreamContent(file);
                _file.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                string _fileName = (file as System.IO.FileStream)?.Name ?? "unknown";
                
                ContentDispositionHeaderValue _contentDispositionHeaderValue = new ContentDispositionHeaderValue("form-data");
                _contentDispositionHeaderValue.Name = "file";

                _contentDispositionHeaderValue.FileName = _fileName;
   
                _file.Headers.ContentDisposition = _contentDispositionHeaderValue;
                _multiPartContent.Add(_file, "file");

Is that something you could fix? Or at least havean overload, one taking a FileStream and another taking a Stream and a Filename?

**Comment by fearthecowboy** [comment]
Hmm. That might be a good idea.  

Gimme a bit and I'll see if I we can get something working.

**Comment by nielsvoigt** [comment]
While you are at it, at the same spot that sets the FileName property, could you check whether the filename contains non-ASCII characters and use _contentDispositionHeaderValue.FileNameStar instead?

[MSDN](https://msdn.microsoft.com/en-US/library/system.net.http.headers.contentdispositionheadervalue.filenamestar(v=vs.110).aspx):

> The FileNameStar property uses IETF RFC 5987 encoding. The FileName and FileNameStar properties differ only in thatFileNameStar uses the encoding defined in IETF RFC 5987, allowing the use of characters not present in the ISO-8859-1 character set.

The example is simple and effective, maybe some codepage/encoding magic would be more correct for ISO-8859-1 above 127.

for example:

```
      if(_fileName.Any(c => c > 127))
      {
           // non ASCII chars detected, need UTF encoding:
            _contentDispositionHeaderValue.FileNameStar = _fileName;
      }
      else
      {
           // ASCII only
            _contentDispositionHeaderValue.FileName = _fileName;
      } 
```

**Comment by skrusty** [comment]
Yeah, we see the same as @nielsvoigt - it posts a blank/empty request (169 bytes).

The only way round it we have at the moment is to create a filestream and pass it in, which is not great.

Thanks in advance.

**Comment by fearthecowboy** [comment]
I'm publishing a one-off build for you to test with:

use 
``` bash
# clear old versions of plugins
autorest --reset 

# run with specific plugin tagged 'stream-filename'
autorest "--use=@microsoft.azure/autorest.csharp@stream-filename" --csharp  <rest of the cmdline...>
```

**Comment by nielsvoigt** [comment]
ok cool, I will try to test it today (my working day starts now, so hopefully I'll test it inside the next 8-9 hours).

**Comment by fearthecowboy** [comment]
I'd rather do the really cheap failover to unicode than bother with `Encoding` at this point. 

I'm pretty sure we're ok with the call.

I'll switch the `Any()` call to an explicit call to `Enumerable.Any` and let the simplifier sort it out. 

Can you try the latest build for this once more and see if it works without tweaking the usings?

``` bash
# clear old versions of plugins again
autorest --reset 

# run with specific plugin tagged 'stream-filename' (latest is 2.2.60)
autorest "--use=@microsoft.azure/autorest.csharp@stream-filename" --csharp  <rest of the cmdline...>
```

**Comment by nielsvoigt** [comment]
I just re-generated my client. Compiles fine with the explicit `Any` call now. Works well, my file gets transferred using both FileStream and MemoryStream to an IFormFile on the controller side. 

**ThumbsUp**

Thanks!

**Comment by skrusty** [comment]
I will try this later, but previous comments seem to suggest this is a good fix. Thank you

**Comment by OliverCole** [comment]
While we're here, this isn't super great:

    _contentDispositionHeaderValue.FileName = _fileContentAsFileStream.Name

Because `Name` is the full file path, resulting in leaking the full path over HTTP to the server (potentially including private details). Plus you may have to manipulate the path on the server to get just the filename, and you might not be doing that on Windows/.NET.

**Comment by skrusty** [comment]
Hi,

I can't get autorest to build the client using the new code generator extension. I am using a readme.md file, and adding the --use via the command line. Should this work? The output seems to be the same as stock and the output indicates it's using the local copy of the generator extension.

Thanks in advance for any assistance

**Comment by fearthecowboy** [comment]
It should work -- add `--debug --verbose` to see what it's doing. 

What's full cmdline/readme.md look like?

**Comment by skrusty** [comment]
Hi @fearthecowboy 

readme.md, command-line and output located here:

https://gist.github.com/skrusty/429e17a151071c0d5b1cc478b24cc274

Thanks in advance.

**Comment by fearthecowboy** [comment]
You are using a really old version of autorest.

upgrade: 

``` batch
npm install -g autorest
autorest --reset 
autorest --info
```

**Comment by skrusty** [comment]
OK, so that seems to be using the new extension now, but generates the following code, which isn't valid.

```c#
if(file != null)
            {
                values.Add(new KeyValuePair<string,string>("file", file));
            }
```

**Comment by anderslm** [comment]
@henning-krause check that you use "multipart/form-data" in your Consumes, eg.
`var consumes = operation.Consumes;
consumes.Clear();
consumes.Add("multipart/form-data");`

If you use "application/form-data" or something else that isn't supported, then that error can happen.

**Comment by henning-krause** [comment]
@anderslm, thanks that did the trick.

**Comment by mahwishs** [comment]
So i tried the autorest "--use=@microsoft.azure/autorest.csharp@stream-filename" --csharp  <rest of the cmdline...>

I am getting syntax error on "file as FileStream" this should have been "backgroundImage  as FileStream"

 if (backgroundImage != null)
            {
                StreamContent _backgroundImage = new StreamContent(backgroundImage);
                _backgroundImage.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                ContentDispositionHeaderValue _contentDispositionHeaderValue = new ContentDispositionHeaderValue("form-data");
                _contentDispositionHeaderValue.Name = "backgroundImage";
                // get filename from stream if it's a file otherwise, just use  'unknown'
                **var _fileName = (file as FileStream)?.Name ?? "unknown";**

I am using the latest version '@microsoft.azure/autorest.csharp' (stream-filename->2.2.60)

what am i missing?

**Comment by skrusty** [comment]
Thanks, tested this and it's fixed our issues!

**Comment by OzBob** [comment]
@mahwishs I would suggest raising yours as a separate issue.

From https://stackoverflow.com/a/41204533/1586498 says

"According to RFC7578 Section 4.3:
[RFC2388] suggested that multiple files for a single form field be transmitted using a nested "multipart/mixed" part ... is deprecated.
To match widely deployed implementations, multiple files MUST be sent by supplying each file in a separate part but all with the same 'name' parameter."

To adhere to the RFC standards implementation, "same 'name' parameter", means that on your API definition you need to accept multiple files as one parameter.

**Comment by eskaufel** [comment]
> moving

Any update on this?
